webapp/lib/controller.dart

```diff
@@ -267,48 +277,86 @@ bool isProjectStale(model.NeedsReplyData projectData) {
+    return SYSTEM_METRICS_ROOT_COLLECTION_KEY;
+  } else {
+    throw new model.ModelDoesNotExistException('Model does not exist');
+  }
+}
+
+void setupWatchdogTimer<T>(T latestData, [bool stale = false]) {
+  var data;
+  if (latestData is model.NeedsReplyData) {
+    data = latestData as model.NeedsReplyData;
+  } else if (latestData is model.SystemMetricsData) {
+    data = latestData as model.SystemMetricsData;
+  }
+
+  watchdogTimers[getWatchdogTimer(data)]?.cancel();
 
   if (stale) {
-    projectTimers[selectedProject] = null;
+    watchdogTimers[getWatchdogTimer(data)] = null;
   } else {
-    var timeToExecute =  projectData.datetime.add(Duration(hours: 2));
+    var timeToExecute =  data.datetime.add(Duration(hours: 2));
     var now = new DateTime.now();
     var duration = timeToExecute.difference(now);
     var timer = new Timer(duration, () {
-      view.contentView.stale = true;
+      if (data?.project == selectedProject) {
+        view.contentView.setStale(NEEDS_REPLY_METRICS_COLLECTION_KEY, true);
+      }
+      view.contentView.setStale(SYSTEM_METRICS_ROOT_COLLECTION_KEY, true);
     });
-    projectTimers[selectedProject] = timer;
+    watchdogTimers[getWatchdogTimer(data)] = timer;
   }
 }
 
 void checkNeedsReplyMetricsStale(List<model.NeedsReplyData> updatedData) {
   if (updatedData.isEmpty) {
-    var selectedProjectTimer = projectTimers[selectedProject];
-    selectedProjectTimer?.cancel();
-    view.contentView.stale = false;
+    var selectedTimer = watchdogTimers[selectedProject];
+    selectedTimer?.cancel();
+    view.contentView.setStale(NEEDS_REPLY_METRICS_COLLECTION_KEY, false);
     return;
   }
 
   updatedData.sort((d1, d2) => d1.datetime.compareTo(d2.datetime));
 
-  var latestProjectData = updatedData.last ?? null;
+  var latestData = updatedData.last ?? null;
 
-  if (latestProjectData != null) {
+  if (latestData != null) {
 
-      if (isProjectStale(latestProjectData)) {
-        setupProjectTimer(latestProjectData, true);
+      if (isDataStale(latestData)) {
+        setupWatchdogTimer(latestData, true);
       } else {
-        setupProjectTimer(latestProjectData);
+        setupWatchdogTimer(latestData);
       }
     }
 
-  var selectedProjectTimer = projectTimers[selectedProject];
-  if (selectedProjectTimer != null && selectedProjectTimer.isActive) {
-    view.contentView.stale = false;
+  var selectedTimer = watchdogTimers[selectedProject];
+  if (selectedTimer != null && selectedTimer.isActive) {
+    view.contentView.setStale(NEEDS_REPLY_METRICS_COLLECTION_KEY, false);
+  } else {
+    view.contentView.setStale(NEEDS_REPLY_METRICS_COLLECTION_KEY, true);
+  }
+}
+
+void checkSystemMetricsStale(List<model.SystemMetricsData> updatedData) {
+  updatedData.sort((d1, d2) => d1.datetime.compareTo(d2.datetime));
+  var latestData = updatedData.last;
+
+  if (isDataStale(latestData)) {
+    setupWatchdogTimer(latestData, true);
+  } else {
+    setupWatchdogTimer(latestData);
+  }
```

Comment:
This can be written in one line:

```suggestion
  setupWatchdogTimer(latestData, isDataStale(latestData));
```

Reply:
Good idea. Refactored it like this.